# <font color=blue>Assignments for "Understanding The Relationship"</font>

To close out this lesson, you're going to do three assignments. For the first assignment, you'll write up a short answer to a question in a Gist file.  For the second two assignments, you'll do your work in Jupyter notebooks, and you should link to those notebooks in the same Gist file.

Please submit a single Gist file containing the answer to first assignment, plus links for second two.

## 1. Interpretion and signficance

Suppose that we would like to know how much families in the US are spending on recreation annually. We estimated the following model:

$$ expenditure = 873 + 0.0012annual\_income + 0.00002annual\_income^2 - 223.57have\_kids $$

*expenditure* is the annual spending on recreation in US dollars, *annual_income* is the annual income in US dollars and *have_kids* is a dummy variable indicating the families with children. Interpret the estimated coefficients. What additional statistics should be given in order to make sure that your interpretations make sense statistically. Write up your answer and save in a Gist. 

- In this equation expenditure is dependent to annual income and having kids. There is a direct proportion between expenditure and income. So, income has a positive affect on the expenditures. On the other hand, increase in the number of childiren has a strong negative impact on expenditures.

## 2. Weather model

In this exercise, you'll work with the historical temperature data from the previous lesson. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
from sklearn import linear_model
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

- Build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. Are the estimated coefficients statistically significant? Are the signs of the estimated coefficients in line with your previous expectations? Interpret the estimated coefficients. What are the relations between the target and the explanatory variables? 

In [9]:
weather_df['difference'] = weather_df['Temperature (C)']-weather_df['Apparent Temperature (C)']

In [16]:
y = weather_df[['difference']]
X = weather_df[['Humidity','Wind Speed (km/h)']]
lrm = linear_model.LinearRegression()
lrm.fit(X, y)

print('\nIntercept: \n', lrm.intercept_)
print('\nCoefficients: \n', lrm.coef_)


Intercept: 
 [-2.43810542]

Coefficients: 
 [[3.02918594 0.11929075]]


In [19]:
X = sm.add_constant(X)

In [20]:
results = sm.OLS(y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             difference   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Thu, 06 May 2021   Prob (F-statistic):               0.00
Time:                        19:20:34   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- P-Values for both coefficients are low. Statistically significant.

- On the previous case Humidity and wind speed coefficients were -3.24 and -2.01;
they are 3.02 and 0.11 right now. humidity changed to negative and wind speed reduced to almost 0

- Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Are the coefficients statistically significant? Did the signs of the estimated coefficients for *humidity* and *windspeed* change? Interpret the estimated coefficients.

In [17]:
yy = weather_df[['difference']]

weather_df["Humidity * Wind Speed"] = weather_df.Humidity * weather_df["Wind Speed (km/h)"]

XX = weather_df[["Humidity", "Wind Speed (km/h)", "Humidity * Wind Speed"]]
lrm = linear_model.LinearRegression()
lrm.fit(XX, yy)

print('\nIntercept: \n', lrm.intercept_)
print('\nCoefficients: \n', lrm.coef_)


Intercept: 
 [-0.08393631]

Coefficients: 
 [[-0.17751219 -0.09048213  0.29711946]]


In [22]:
XX = sm.add_constant(XX)
new_results = sm.OLS(yy, XX).fit()
new_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             difference   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Thu, 06 May 2021   Prob (F-statistic):               0.00
Time:                        19:21:37   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                    -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity                 -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)        -0.0905      0.002    -36.797      0.000      -0.095      -0.086
Humidity * Wind Speed     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- P-Value has slightly changed! All coefficients got closer to zero. Both signs changed to negative.

##  3. House prices model

In this exercise, you'll interpret your house prices model. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [23]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

- Run your house prices model again and interpret the results. Which features are statistically significant and which are not?

In [33]:
y1 = house_prices_df["SalePrice"]
X1 = house_prices_df[['OverallQual', 'TotalBsmtSF','1stFlrSF','GrLivArea','GarageCars','GarageArea','YrSold']]

X1 = sm.add_constant(X1)
results1 = sm.OLS(y1, X1).fit()
results1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.762
Model:                            OLS   Adj. R-squared:                  0.761
Method:                 Least Squares   F-statistic:                     663.8
Date:                Sun, 09 May 2021   Prob (F-statistic):               0.00
Time:                        15:09:05   Log-Likelihood:                -17496.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1452   BIC:                         3.505e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -3.642e+05   1.54e+06     -0.237      0.813   -3.38e+06    2.66e+06
OverallQual    2.4e+04   1083.767     22.141      0.000    2.19e+04    2.61e+04
TotalBsmtSF    24.3949      4.319      5.648      0.000      15.922      32.867
1stFlrSF       11.1760      5.034      2.220      0.027       1.301      21.051
GrLivArea      43.1346      2.681     16.089      0.000      37.875      48.394
GarageCars   1.453e+04   3020.839      4.810      0.000    8604.032    2.05e+04
GarageArea     15.6356     10.480      1.492      0.136      -4.922      36.193
YrSold        130.2681    766.892      0.170      0.865   -1374.067    1634.603
==============================================================================
Omnibus:                      431.830   Durbin-Watson:                   1.976
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            39204.109
Skew:                          -0.314   Prob(JB):                         0.00
Kurtosis:                      28.378   Cond. No.                     4.61e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.61e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- Now, exclude the insignificant features from your model. Did anything change?

In [36]:
# '1stFlrSF', GarageArea','YrSold' variables are insignificant features.
df_prices = house_prices_df.filter(['OverallQual', 'TotalBsmtSF','GrLivArea','GarageCars','SalePrice'], axis=1)

y1 = df_prices["SalePrice"]
X1 = df_prices[['OverallQual', 'TotalBsmtSF','GrLivArea','GarageCars']]

X1 = sm.add_constant(X1)
results1 = sm.OLS(y1, X1).fit()
results1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.761
Model:                            OLS   Adj. R-squared:                  0.760
Method:                 Least Squares   F-statistic:                     1156.
Date:                Sun, 09 May 2021   Prob (F-statistic):               0.00
Time:                        15:17:48   Log-Likelihood:                -17500.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1455   BIC:                         3.504e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -9.925e+04   4639.866    -21.390      0.000   -1.08e+05   -9.01e+04
OverallQual  2.357e+04   1072.465     21.980      0.000    2.15e+04    2.57e+04
TotalBsmtSF    32.5203      2.838     11.459      0.000      26.953      38.087
GrLivArea      45.6432      2.484     18.376      0.000      40.771      50.515
GarageCars   1.858e+04   1747.412     10.634      0.000    1.52e+04     2.2e+04
==============================================================================
Omnibus:                      421.692   Durbin-Watson:                   1.971
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            36798.858
Skew:                          -0.273   Prob(JB):                         0.00
Kurtosis:                      27.589   Cond. No.                     8.95e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.95e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- Interpret the statistically significant coefficients by quantifying their relations with the house prices. Which features have more prominent effect on the house prices?

OverallQual, TotalBsmtS, GrLivArea, GarageCars have more prominent effect on the modal because t-scores are 0.

YrSold t-test score is higher than 0,05 which means it has not an impact on our modal. I also removed 1stFlrSF, GarageArea

- Do the results sound reasonable to you? If not, try to explain the potential reasons.

R-squared: R-squared value between 70-90% is ideal. In the modal is 76% which is ok.


Adjusted R-squared: The adjusted R-squared increases when the new term improves the model more than would be expected by chance. It decreases when a predictor improves the model by less than expected. However, Adj R-squared decreased %0,1 percent which is very low difference.
    
    
F-Statistic: was 663.8 sharply increased to 1156. In other words, our new F value in a test is larger than your F critical value, we can reject the null hypothesis. 